File: src/typescript/src/pipeline_services/core/pipeline_router.ts
import { PipelineProvider } from "./pipeline_provider.js";

export class PipelineRouter<EventType, ContextType = unknown> {
  constructor(
    private readonly route: (
      event: EventType,
    ) => PipelineProvider<ContextType>,
  ) {
    if (typeof route !== "function") {
      throw new TypeError("route must be callable");
    }
  }

  getPipelineProvider(
    event: EventType,
  ): PipelineProvider<ContextType> {
    const provider = this.route(event);
    if (!(provider instanceof PipelineProvider)) {
      throw new TypeError("route returned an invalid PipelineProvider");
    }
    return provider;
  }

  get_pipeline_provider(
    event: EventType,
  ): PipelineProvider<ContextType> {
    return this.getPipelineProvider(event);
  }
}

File: src/Python/pipeline_services/core/pipeline_router.py
from __future__ import annotations

from typing import Callable, Generic, TypeVar

from .pipeline_provider import PipelineProvider

EventType = TypeVar("EventType")
ContextType = TypeVar("ContextType")


class PipelineRouter(Generic[EventType, ContextType]):
    """Selects a PipelineProvider for an event without changing provider lifecycle."""

    def __init__(
        self,
        route: Callable[[EventType], PipelineProvider[ContextType]],
    ) -> None:
        if not callable(route):
            raise TypeError("route must be callable")
        self._route = route

    def get_pipeline_provider(
        self,
        event: EventType,
    ) -> PipelineProvider[ContextType]:
        provider = self._route(event)
        if provider is None:
            raise ValueError("route returned None")
        return provider
